File: backend/src/bond_monitor/dev/overrides.py
"""Dev-only notification overrides for local manual testing."""

from __future__ import annotations

import json
import logging
from dataclasses import dataclass
from datetime import date, timedelta
from pathlib import Path
from typing import Any

from bond_monitor.domain.bonds.models import BondRecord
from bond_monitor.domain.portfolio.models import Portfolio, PortfolioPosition
from bond_monitor.domain.portfolio.position_factory import sync_put_offer_from_bond
from bond_monitor.domain.portfolio.risk_monitor import RiskSnapshot
from bond_monitor.infrastructure.paths import get_cache_dir
from bond_monitor.interfaces.config import get_settings

logger = logging.getLogger(__name__)

DEV_OVERRIDES_FILENAME = "dev_notification_overrides.json"


@dataclass(frozen=True)
class DevPutOfferOverride:
    offer_date: date
    submission_start: date
    submission_end: date
    offer_price_pct: float


@dataclass(frozen=True)
class DevOverrides:
    portfolio_id: str
    put_offers: dict[str, DevPutOfferOverride]
    risk_baselines: dict[str, RiskSnapshot]
    bond_risk: dict[str, dict[str, Any]]


def get_dev_overrides_path() -> Path:
    return get_cache_dir() / DEV_OVERRIDES_FILENAME


def notifications_dev_enabled() -> bool:
    return get_settings().notifications_dev


def _parse_date(value: object) -> date | None:
    if not value or not isinstance(value, str):
        return None
    return date.fromisoformat(value)


def _risk_snapshot_from_dict(data: dict[str, Any]) -> RiskSnapshot:
    return RiskSnapshot(
        has_default=bool(data.get("has_default", False)),
        has_technical_default=bool(data.get("has_technical_default", False)),
        credit_rating=(
            str(data["credit_rating"]) if data.get("credit_rating") is not None else None
        ),
    )


def load_dev_overrides(
    path: Path | None = None,
    *,
    portfolio_id: str | None = None,
) -> DevOverrides | None:
    overrides_path = path or get_dev_overrides_path()
    if not overrides_path.exists():
        return None
    try:
        with overrides_path.open(encoding="utf-8") as fh:
            raw = json.load(fh)
    except (OSError, json.JSONDecodeError):
        logger.warning("Dev overrides unreadable at %s", overrides_path, exc_info=True)
        return None
    if not isinstance(raw, dict):
        return None

    file_portfolio_id = str(raw.get("portfolio_id", ""))
    if portfolio_id is not None and file_portfolio_id != portfolio_id:
        return None

    put_offers: dict[str, DevPutOfferOverride] = {}
    for isin, entry in (raw.get("put_offers") or {}).items():
        if not isinstance(entry, dict):
            continue
        offer_date = _parse_date(entry.get("offer_date"))
        submission_start = _parse_date(entry.get("submission_start"))
        submission_end = _parse_date(entry.get("submission_end"))
        if offer_date is None or submission_start is None or submission_end is None:
            continue
        put_offers[str(isin)] = DevPutOfferOverride(
            offer_date=offer_date,
            submission_start=submission_start,
            submission_end=submission_end,
            offer_price_pct=float(entry.get("offer_price_pct", 100.0)),
        )

    risk_baselines = {
        str(isin): _risk_snapshot_from_dict(entry)
        for isin, entry in (raw.get("risk_baselines") or {}).items()
        if isinstance(entry, dict)
    }
    bond_risk = {
        str(isin): entry
        for isin, entry in (raw.get("bond_risk") or {}).items()
        if isinstance(entry, dict)
    }

    if not file_portfolio_id:
        return None

    return DevOverrides(
        portfolio_id=file_portfolio_id,
        put_offers=put_offers,
        risk_baselines=risk_baselines,
        bond_risk=bond_risk,
    )


def save_dev_overrides(path: Path, payload: dict[str, Any]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.with_suffix(".json.tmp")
    with tmp.open("w", encoding="utf-8") as fh:
        json.dump(payload, fh, ensure_ascii=False, indent=2)
    tmp.replace(path)


def build_put_offer_overrides(
    *,
    portfolio_id: str,
    isin: str,
    today: date | None = None,
) -> dict[str, Any]:
    ref = today or date.today()
    offer_date = ref + timedelta(days=10)
    return {
        "portfolio_id": portfolio_id,
        "put_offers": {
            isin: {
                "offer_date": offer_date.isoformat(),
                "submission_start": (ref - timedelta(days=1)).isoformat(),
                "submission_end": (ref + timedelta(days=7)).isoformat(),
                "offer_price_pct": 100.0,
            }
        },
        "risk_baselines": {},
        "bond_risk": {},
    }


def build_risk_default_overrides(*, portfolio_id: str, isin: str) -> dict[str, Any]:
    return {
        "portfolio_id": portfolio_id,
        "put_offers": {},
        "risk_baselines": {
            isin: {
                "has_default": False,
                "has_technical_default": False,
                "credit_rating": "ruBBB",
            }
        },
        "bond_risk": {
            isin: {
                "has_default": True,
                "has_technical_default": False,
            }
        },
    }


def build_risk_downgrade_overrides(*, portfolio_id: str, isin: str) -> dict[str, Any]:
    return {
        "portfolio_id": portfolio_id,
        "put_offers": {},
        "risk_baselines": {
            isin: {
                "has_default": False,
                "has_technical_default": False,
                "credit_rating": "ruBBB-",
            }
        },
        "bond_risk": {
            isin: {
                "credit_rating": "ruBB+",
            }
        },
    }


def _apply_bond_risk_patch(bond: BondRecord, patch: dict[str, Any]) -> None:
    if "credit_rating" in patch:
        bond.credit_rating = (
            str(patch["credit_rating"]) if patch["credit_rating"] is not None else None
        )
    if "has_default" in patch:
        bond.has_default = bool(patch["has_default"])
    if "has_technical_default" in patch:
        bond.has_technical_default = bool(patch["has_technical_default"])


def _apply_put_offer_patch(bond: BondRecord, schedule: DevPutOfferOverride, ref_date: date) -> None:
    bond.offer_date = schedule.offer_date
    bond.offer_submission_start = schedule.submission_start
    bond.offer_submission_end = schedule.submission_end
    bond.offer_price_pct = schedule.offer_price_pct
    dates = [d for d in (bond.maturity_date, bond.offer_date) if d is not None and d >= ref_date]
    if dates:
        bond.effective_date = min(dates)
        bond.days_to_maturity = (bond.effective_date - ref_date).days


def apply_dev_notification_overrides(
    portfolio: Portfolio,
    universe: list[BondRecord],
    positions: list[PortfolioPosition],
    *,
    portfolio_id: str,
    path: Path | None = None,
    today: date | None = None,
) -> bool:
    """Apply dev overrides when file matches portfolio_id. Returns True if applied."""
    overrides = load_dev_overrides(path, portfolio_id=portfolio_id)
    if overrides is None:
        return False

    ref_date = today or date.today()
    universe_by_isin = {bond.isin: bond for bond in universe}

    for isin, patch in overrides.bond_risk.items():
        bond = universe_by_isin.get(isin)
        if bond is not None:
            _apply_bond_risk_patch(bond, patch)

    for isin, baseline in overrides.risk_baselines.items():
        portfolio.risk_baselines[isin] = baseline

    for isin, schedule in overrides.put_offers.items():
        bond = universe_by_isin.get(isin)
        if bond is None:
            continue
        _apply_put_offer_patch(bond, schedule, ref_date)
        for position in positions:
            if position.isin == isin:
                sync_put_offer_from_bond(position, bond)

    logger.info(
        "Applied dev notification overrides for portfolio %s (%d put, %d risk)",
        portfolio_id,
        len(overrides.put_offers),
        len(overrides.bond_risk),
    )
    return True

File: backend/internal/dev/notifications/overrides.go
package notifications

import (
	"encoding/json"
	"os"
	"path/filepath"
	"time"

	"github.com/tonatos/instrumenta/backend/internal/domain/bonds"
	"github.com/tonatos/instrumenta/backend/internal/domain/portfolio"
	"github.com/tonatos/instrumenta/backend/internal/domain/shared"
	"github.com/tonatos/instrumenta/backend/internal/infrastructure/paths"
)

const devOverridesFilename = "dev_notification_overrides.json"

// DevOverridesPath returns the default path for dev notification overrides.
func DevOverridesPath() string {
	return filepath.Join(paths.CacheDir(), devOverridesFilename)
}

type devPutOfferOverride struct {
	OfferDate        string  `json:"offer_date"`
	SubmissionStart  string  `json:"submission_start"`
	SubmissionEnd    string  `json:"submission_end"`
	OfferPricePct    float64 `json:"offer_price_pct"`
}

type devOverridesFile struct {
	PortfolioID   string                           `json:"portfolio_id"`
	PutOffers     map[string]devPutOfferOverride   `json:"put_offers"`
	RiskBaselines map[string]riskSnapshotJSON      `json:"risk_baselines"`
	BondRisk      map[string]map[string]any        `json:"bond_risk"`
}

type riskSnapshotJSON struct {
	HasDefault          bool    `json:"has_default"`
	HasTechnicalDefault bool    `json:"has_technical_default"`
	CreditRating        *string `json:"credit_rating"`
}

type DevOverrides struct {
	PortfolioID   string
	PutOffers     map[string]devPutOfferOverride
	RiskBaselines map[string]portfolio.RiskSnapshot
	BondRisk      map[string]map[string]any
}

// BuildPutOfferOverrides returns fake put-offer data with an open submission window.
func BuildPutOfferOverrides(portfolioID, isin string, today time.Time) map[string]any {
	ref := shared.DateOnly(today)
	offerDate := shared.AddDays(ref, 10)
	return map[string]any{
		"portfolio_id": portfolioID,
		"put_offers": map[string]any{
			isin: map[string]any{
				"offer_date":        shared.FormatISODate(offerDate),
				"submission_start":  shared.FormatISODate(shared.AddDays(ref, -1)),
				"submission_end":    shared.FormatISODate(shared.AddDays(ref, 7)),
				"offer_price_pct":   100.0,
			},
		},
		"risk_baselines": map[string]any{},
		"bond_risk":      map[string]any{},
	}
}

// BuildRiskDefaultOverrides returns fake data that triggers a critical default escalation.
func BuildRiskDefaultOverrides(portfolioID, isin string) map[string]any {
	rating := "ruBBB"
	return map[string]any{
		"portfolio_id": portfolioID,
		"put_offers":   map[string]any{},
		"risk_baselines": map[string]any{
			isin: map[string]any{
				"has_default":           false,
				"has_technical_default": false,
				"credit_rating":         rating,
			},
		},
		"bond_risk": map[string]any{
			isin: map[string]any{
				"has_default":           true,
				"has_technical_default": false,
			},
		},
	}
}

// BuildRiskDowngradeOverrides returns fake data that triggers a rating downgrade escalation.
func BuildRiskDowngradeOverrides(portfolioID, isin string) map[string]any {
	baseline := "ruBBB-"
	current := "ruBB+"
	return map[string]any{
		"portfolio_id": portfolioID,
		"put_offers":   map[string]any{},
		"risk_baselines": map[string]any{
			isin: map[string]any{
				"has_default":           false,
				"has_technical_default": false,
				"credit_rating":         baseline,
			},
		},
		"bond_risk": map[string]any{
			isin: map[string]any{
				"credit_rating": current,
			},
		},
	}
}

// LoadDevOverrides reads overrides for portfolioID. Returns nil if missing or mismatched.
func LoadDevOverrides(path, portfolioID string) *DevOverrides {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil
	}
	var raw devOverridesFile
	if err := json.Unmarshal(data, &raw); err != nil {
		return nil
	}
	if raw.PortfolioID == "" || raw.PortfolioID != portfolioID {
		return nil
	}
	result := &DevOverrides{
		PortfolioID:   raw.PortfolioID,
		PutOffers:     raw.PutOffers,
		RiskBaselines: make(map[string]portfolio.RiskSnapshot, len(raw.RiskBaselines)),
		BondRisk:      raw.BondRisk,
	}
	for isin, entry := range raw.RiskBaselines {
		result.RiskBaselines[isin] = riskSnapshotFromJSON(entry)
	}
	if result.PutOffers == nil {
		result.PutOffers = map[string]devPutOfferOverride{}
	}
	if result.BondRisk == nil {
		result.BondRisk = map[string]map[string]any{}
	}
	return result
}

// SaveDevOverrides atomically writes overrides JSON.
func SaveDevOverrides(path string, payload map[string]any) error {
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	data, err := json.MarshalIndent(payload, "", "  ")
	if err != nil {
		return err
	}
	tmp := path + ".tmp"
	if err := os.WriteFile(tmp, data, 0o644); err != nil {
		return err
	}
	return os.Rename(tmp, path)
}

// ApplyDevNotificationOverrides patches universe, portfolio baselines, and positions in memory.
// Returns true when overrides were applied for portfolioID.
func ApplyDevNotificationOverrides(
	p *portfolio.Portfolio,
	universe []bonds.BondRecord,
	positions []portfolio.PortfolioPosition,
	portfolioID, path string,
	today time.Time,
) bool {
	overrides := LoadDevOverrides(path, portfolioID)
	if overrides == nil {
		return false
	}
	refDate := shared.DateOnly(today)
	universeByISIN := make(map[string]*bonds.BondRecord, len(universe))
	for i := range universe {
		universeByISIN[universe[i].ISIN] = &universe[i]
	}

	for isin, patch := range overrides.BondRisk {
		if bond, ok := universeByISIN[isin]; ok {
			applyBondRiskPatch(bond, patch)
		}
	}
	for isin, baseline := range overrides.RiskBaselines {
		if p.RiskBaselines == nil {
			p.RiskBaselines = make(map[string]portfolio.RiskSnapshot)
		}
		p.RiskBaselines[isin] = baseline
	}
	for isin, schedule := range overrides.PutOffers {
		bond, ok := universeByISIN[isin]
		if !ok {
			continue
		}
		applyPutOfferPatch(bond, schedule, refDate)
		for i := range positions {
			if positions[i].ISIN == isin {
				portfolio.SyncPutOfferFromBond(&positions[i], *bond)
			}
		}
	}
	return true
}

func riskSnapshotFromJSON(entry riskSnapshotJSON) portfolio.RiskSnapshot {
	return portfolio.RiskSnapshot{
		HasDefault:          entry.HasDefault,
		HasTechnicalDefault: entry.HasTechnicalDefault,
		CreditRating:        entry.CreditRating,
	}
}

func applyBondRiskPatch(bond *bonds.BondRecord, patch map[string]any) {
	if v, ok := patch["credit_rating"]; ok {
		if v == nil {
			bond.CreditRating = nil
		} else if s, ok := v.(string); ok {
			bond.CreditRating = &s
		}
	}
	if v, ok := patch["has_default"]; ok {
		if b, ok := v.(bool); ok {
			bond.HasDefault = b
		}
	}
	if v, ok := patch["has_technical_default"]; ok {
		if b, ok := v.(bool); ok {
			bond.HasTechnicalDefault = b
		}
	}
}

func applyPutOfferPatch(bond *bonds.BondRecord, schedule devPutOfferOverride, refDate time.Time) {
	offerDate := parseDate(schedule.OfferDate)
	submissionStart := parseDate(schedule.SubmissionStart)
	submissionEnd := parseDate(schedule.SubmissionEnd)
	if offerDate == nil || submissionStart == nil || submissionEnd == nil {
		return
	}
	bond.OfferDate = offerDate
	bond.OfferSubmissionStart = submissionStart
	bond.OfferSubmissionEnd = submissionEnd
	price := schedule.OfferPricePct
	if price == 0 {
		price = 100
	}
	bond.OfferPricePct = &price

	var dates []time.Time
	if bond.MaturityDate != nil && !bond.MaturityDate.Before(refDate) {
		dates = append(dates, *bond.MaturityDate)
	}
	if !offerDate.Before(refDate) {
		dates = append(dates, *offerDate)
	}
	if len(dates) > 0 {
		effective := dates[0]
		for _, d := range dates[1:] {
			if d.Before(effective) {
				effective = d
			}
		}
		bond.EffectiveDate = &effective
		days := shared.DaysBetween(refDate, effective)
		bond.DaysToMaturity = &days
	}
}

func parseDate(value string) *time.Time {
	if value == "" {
		return nil
	}
	t, err := time.Parse("2006-01-02", value)
	if err != nil {
		return nil
	}
	d := shared.DateOnly(t)
	return &d
}
